articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,230 @@
+---
+layout: default
+title: Proposal for Implementation of Real-time Systems in ROS 2
+permalink: articles/realtime_proposal.html
+abstract: This article proposes an approach to the real-time performance requirement in ROS 2.
+published: true 
+author: Jackie Kay
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Requirements and Implementation of Real-Time Systems
+
+### System requirements
+
+The requirements of a real-time system varies depending on the use case.
+At its core, the real-time requirement of a system has two components:
+
+- Latency
+
+  - Update period
+  - Determinism (measured by jitter)
+
+- Failure mode
+
+  - How to react to a missed deadline
+
+Language referring to hard/soft/firm real-time systems generally refers to the failure mode.
+A hard real-time system treats a missed deadline as a system failure.
+A soft real-time system tries to meet deadlines but does not fail when a deadline is missed.
+A firm real-time system discards computations made by missed deadlines and may degrade its performance requirement in order to adapt to a missed deadline.
+
+Though in this document we consider the failure mode orthogonal to latency characteristics, hard real-time systems are usually designed to be safety critical and therefore require high determinism (low jitter).
+
+### Implementation choices: computer architecture and OS
+
+Processors vary in capability, number of cores, memory size, and instruction set architectures, some which are better suited to real-time performance than others.
+The selection of computer architecture may be due to cost, size, or energy efficiency, which could set a constraint for the rest of the system.
+A single robot may contain multiple computers with a variety of processor types, from the main control board to sensors.
+
+Another major consideration when designing a real-time system is the choice of operating system.
+A true "hard" real-time operating system is fully deterministic.
+"Soft" real-time operating systems are designed to meet most deadlines deterministically, but may still have an element of nondeterminism.
+Most conventional user-facing operating systems are not designed with determinism or fairness as the first priority, often because they have to support graphical displays and interactivity.
+Here are some features critical to deterministic performance that are often exposed to the user in an RTOS, but not in a conventional OS (1).
```

Comment:
We could either use at least a link or don't put the URLs at the bottom but link them directly inline?